diff -urN claws-mail-3.7.4-old/src/gtk/filesel.c claws-mail-3.7.4-new/src/gtk/filesel.c
--- claws-mail-3.7.4-old/src/gtk/filesel.c	2009-01-09 18:47:16.000000000 +0100
+++ claws-mail-3.7.4-new/src/gtk/filesel.c	2010-01-18 20:51:06.000000000 +0100
@@ -226,13 +226,17 @@
 	}
 
 	while (slist) {
-		list = g_list_append(list, slist->data);
+		 gchar *tmp = g_strdup(slist->data);
+
+		if(strncmp(slist->data,"/home/user/Download/Mail/",25)==0){
+			list = g_list_append(list, slist->data);
+		}
 		slist = slist->next;
+		g_free(tmp);
 	}
 	
 	if (slist_orig)
 		g_slist_free(slist_orig);
-	
 	return list;
 }
 
