diff -urN claws-mail-3.7.5/src/compose.c claws-mail-3.7.5-new/src/compose.c
--- claws-mail-3.7.5/src/compose.c	2010-05-27 13:13:26.157738302 +0200
+++ claws-mail-3.7.5-new/src/compose.c	2010-05-27 13:14:42.503530152 +0200
@@ -1308,7 +1308,7 @@
 	cm_return_val_if_fail(msginfo != NULL, NULL);
 
 	list_len = g_slist_length(msginfo_list);
-	
+
 	// Check if we have the right to open a reply compose window
 	refuseLoading=hooks_invoke(COMPOSE_ON_REPLY_MAIL_HOOKLIST, msginfo);
 	if(refuseLoading)
diff -urN claws-mail-3.7.5/src/compose.h claws-mail-3.7.5-new/src/compose.h
--- claws-mail-3.7.5/src/compose.h	2010-05-27 13:13:26.161738571 +0200
+++ claws-mail-3.7.5-new/src/compose.h	2010-05-27 13:13:57.303742364 +0200
@@ -49,6 +49,7 @@
 #define COMPOSE_ON_REPLY_MAIL_HOOKLIST "compose_on_reply_mail_hooklist"
 #define COMPOSE_CHECK_ATTACH_FILES_HOOKLIST "compose_check_attach_files_hooklist"
 #define COMPOSE_CHECK_INSERT_FILES_HOOKLIST "compose_check_insert_files_hooklist"
+#define COMPOSE_ON_REPLY_MAIL_HOOKLIST "compose_on_reply_mail_hooklist"
 
 typedef enum
 {
